packages/application-config/src/substitute-variable-placeholders.ts

```diff
@@ -0,0 +1,154 @@
+  // See https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/test
+  //    As with exec() (or in combination with it), test() called multiple times
+  //    on the same global regular expression instance will advance past the previous match.
+  Boolean(valueOfEnvConfig.match(variableSyntax));
+
+const isEnvVariablePlaceholder = (valueOfPlaceholder: string) =>
+  Boolean(valueOfPlaceholder.match(envRefSyntax));
+
+const isIntlVariablePlaceholder = (valueOfPlaceholder: string) =>
+  Boolean(valueOfPlaceholder.match(intlRefSyntax));
+
+const isFilePathVariablePlaceholder = (valueOfPlaceholder: string) =>
+  Boolean(valueOfPlaceholder.match(filePathRefSyntax));
+
+const substituteEnvVariablePlaceholder = (
+  valueOfPlaceholder: string,
+  matchedString: string,
+  valueOfEnvConfig: string,
+  meta: Meta
+) => {
+  const [, requestedEnvVar] = valueOfPlaceholder.split(':');
+  const hasEnvField = meta.processEnv.hasOwnProperty(requestedEnvVar);
+
+  if (!hasEnvField) {
+    throw new Error(
+      `Missing environment variable '${requestedEnvVar}' specified in config as 'env:${requestedEnvVar}'.`
+    );
+  }
+
+  const escapedMatchedString = matchedString.replace(/[${}:]/g, '\\$&');
+  return valueOfEnvConfig.replace(
+    new RegExp(`(${escapedMatchedString})+`, 'g'),
+    meta.processEnv[requestedEnvVar] as string
+  );
+};
+
+const substituteIntlVariablePlaceholder = (
+  valueOfPlaceholder: string,
+  matchedString: string,
+  valueOfEnvConfig: string,
+  meta: Meta
+) => {
+  const [, locale, requestedIntlMessageId] = valueOfPlaceholder.split(':');
+
+  const translationsFilePath = require.resolve(`./i18n/data/${locale}.json`, {
+    paths: [`${meta.applicationPath}/src`, meta.applicationPath],
+  });
+  const translations: Record<string, string> = require(translationsFilePath);
+
+  const hasIntlMessage = translations.hasOwnProperty(requestedIntlMessageId);
+
+  if (!hasIntlMessage) {
+    throw new Error(
+      `Missing message key '${requestedIntlMessageId}' specified in config as 'intl:${locale}:${requestedIntlMessageId}'.`
+    );
+  }
+
+  const escapedMatchedString = matchedString.replace(/[${}:]/g, '\\$&');
+  return valueOfEnvConfig.replace(
+    new RegExp(`(${escapedMatchedString})+`, 'g'),
+    translations[requestedIntlMessageId]
+  );
+};
+
+const substituteFilePathVariablePlaceholder = (
+  valueOfPlaceholder: string,
+  matchedString: string,
+  valueOfEnvConfig: string,
+  meta: Meta
+) => {
+  const [, filePathOrModule] = valueOfPlaceholder.split(':');
+
+  const content = fs.readFileSync(
+    require.resolve(filePathOrModule, {
+      // Relative paths should be resolved from the application folder.
+      paths: [meta.applicationPath],
+    }),
+    {
+      encoding: 'utf-8',
+    }
+  );
+
+  const escapedMatchedString = matchedString.replace(/[${}:]/g, '\\$&');
+  return valueOfEnvConfig.replace(
+    new RegExp(`(${escapedMatchedString})+`, 'g'),
+    content
+  );
+};
```

Comment:
This resolves the references of file paths, to load and inline the content as a string.